fastjsonschema/generator.py

```diff
@@ -414,3 +470,46 @@ def generate_properties(self):
                         '{}_value'.format(self._variable),
                         '{}.{{{}_key}}'.format(self._variable_name, self._variable),
                     )
+
+    def generate_additional_properties(self):
+        with self.l('if not isinstance({variable}, dict):'):
+            self.l('return {variable}')
+        self.l('{variable}_keys = set({variable}.keys())')
+
+        if 'patternProperties' in self._definition:
+            self.l('pattern_keys = set()')
+            with self.l('for key in {variable}_keys:'):
+                for pattern in self._definition['patternProperties'].keys():
+                    with self.l('if globals()["{}_re"].search(key):', pattern):
+                        self.l('pattern_keys.add(key)')
+            self.l('{variable}_keys -= pattern_keys')
+
+        add_prop_definition = self._definition["additionalProperties"]
+        if add_prop_definition:
+            with self.l('for {variable}_key in {variable}_keys:'):
+                with self.l('if {variable}_key not in "{}":', self._definition.get('properties', [])):
+                    self.l('{variable}_value = {variable}.get({variable}_key)')
+                    self.generate_func_code_block(
+                        add_prop_definition,
+                        '{}_value'.format(self._variable),
+                        '{}.{{{}_key}}'.format(self._variable_name, self._variable),
+                    )
+            if 'default' in add_prop_definition:
+                self.l('else: {variable}["{}"] = {}', key, repr(add_prop_definition['default']))
+
+    def generate_pattern_properties(self):
+        with self.l('if not isinstance({variable}, dict):'):
+            self.l('return {variable}')
+        for pattern, definition in self._definition['patternProperties'].items():
+            self._compile_regexps['{}_re'.format(pattern)] = re.compile(pattern)
```

Comment:
I tried the non-`globals()` index-based approach first and I worried about ordering during key iteration so I refactored to this may-based approach.  Ordering shouldn't be an issue if the `dict` is never mutated but I didn't know how to be sure about that beyond converting the schema definition to an `OrderedDict` and that felt like too much change.

I don't love polluting `globals` but this seemed like the lesser of evils.  A local solution would be faster but I haven't profiled the difference.